File: 78.js
function nextGreater(nums) {
  const n = nums.length;
  const res = Array(n).fill(-1);
  const st = [];

  for (let i = 0; i < n; i++) {
    while (st.length && nums[i] > nums[st[st.length - 1]]) {
      const idx = st.pop();
      res[idx] = nums[i];
    }
    st.push(i);
  }

  return res;
}

// Driver
console.log(nextGreater([2,1,2,4,3]));

File: 78.py
def nextGreater(nums):
    n = len(nums)
    res = [-1] * n
    stack = []  # stores indexes

    for i in range(n):
        while stack and nums[i] > nums[stack[-1]]:
            idx = stack.pop()
            res[idx] = nums[i]
        stack.append(i)

    return res

# Driver
print(nextGreater([2,1,2,4,3]))
